src/Language/Futhark/Interpreter.hs

```diff
@@ -1921,11 +1984,90 @@ initialCtx =
+
+        -- Run the function
+        out <- apply noLoc mempty f v'
+
+        -- Derive the seeds
+        let deriveSeeds j i v = case v of
+              ValuePrim v' -> (i + 1, primToAD v')
+              ValueSeed s@(AD.VjpSeed d tp) ->
+                if d == depth then
+                  case M.lookup j <$> AD.deriveVjp tp (s' !! i) of
+                    Just (Just v') -> (i + 1, v')
+                    _ -> (i + 1, AD.Primal $ AD.valueAsType (AD.value $ AD.primal s) 0)
+                else (i + 1, AD.Primal $ AD.valueAsType (AD.value $ AD.primal s) 0)
+              ValueRecord m -> do
+                let acc = M.mapAccum (deriveSeeds j) i m -- TODO: Fix types
+                (fst acc, foldl (\x y -> fromJust $ adBinOp (AD.OpBin $ P.FAdd Float64) x y) (AD.Primal $ P.FloatValue $ Float64Value 0) $ M.elems $ snd acc)
+              _ -> error "Not implemented (d19h782)" -- TODO: Arrays?
+        
+        let deriveFromV i v = case v of
+              ValuePrim _ -> (i + 1, adToValue $ snd $ deriveSeeds i 0 out)
+              ValueSeed _ -> (i + 1, adToValue $ snd $ deriveSeeds i 0 out)
+              ValueRecord m -> second ValueRecord $ M.mapAccum deriveFromV i m
+              _ -> error "Not implemented (d19h782)" -- TODO: Arrays?
+
+        -- Make sure to return a tuple
+        let k = snd $ deriveFromV 0 v
+        pure $ toTuple [k, k] -- TODO: ???
+
     def "jvp2" = Just $
-      fun3 $
-        \_ _ _ -> bad noLoc mempty "Interpreter does not support autodiff."
+      fun3 $ \f v s -> do
+        -- Get the depth
+        depth <- length <$> stacktrace
+
+        -- Get the seed values
+        let getSeeds v = case v of
+              ValuePrim _ -> [valueToAD v]
+              ValueSeed _ -> [valueToAD v]
+              ValueRecord m -> foldl1 (++) (M.map getSeeds m)
+              _ -> error "Not implemented (d19h45iu782)" -- TODO: Arrays?
+        let s' = getSeeds s
+
+        -- Impregnate the values
+        let addSeeds i v = case v of
+              ValuePrim v' -> (i + 1, ValueSeed $ AD.JvpSeed depth (AD.Primal $ fromJust $ getV v') $ s' !! i)
+              ValueSeed se -> (i + 1, ValueSeed $ AD.JvpSeed depth (AD.Seed se) $ s' !! i)
+              ValueRecord m -> second ValueRecord $ M.mapAccum addSeeds i m
+              _ -> error "Not implemented (d19h45iu782)" -- TODO: Arrays?
+        let v' = snd $ addSeeds 0 v
+
+        -- Run the function
+        out <- apply noLoc mempty f v'
+
+        -- Derive the seeds
+        let deriveSeeds v = case v of
+              ValuePrim v' -> ValuePrim $ putV $ AD.valueAsType (fromJust $ getV v') 0
+              ValueSeed s@(AD.JvpSeed d _ d') ->
+                if d == depth then -- TODO: Fix add op
+                  adToValue d'
+                else ValuePrim $ putV $ AD.valueAsType (AD.value $ AD.primal s) 0
+              ValueRecord m -> ValueRecord $ M.map deriveSeeds m
+              _ -> error "Not implemented (d19hasd782)" -- TODO: Arrays?
```

Comment:
I have a hunch that some functions (like this one) should be factored into their own top-level definitions, to make it easier to understand.